0-common/scala-core/src/main/scala/com.snowplowanalytics.iglu/core/SelfDescribed.scala

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2016 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.iglu
+package core
+
+// Scala
+import scala.language.implicitConversions
+
+/**
+ * Type-class representing essence of Self-describing data
+ * It can be instantiated for any type [[E]] which is potentially
+ * containing its own description (as [[SchemaKey]])
+ *
+ * @tparam E entity type, mostly intended for various JSON ADTs,
+ *           like Json4s, Jackson, circe, Argonaut etc,
+ *           but also can be anything that can bear reference to
+ *           its description like Thrift, Map[String, String] etc
+ */
+trait SelfDescribed[E] {
```

Comment:
Nope, that was not a typo. 
If some type `E` have instance for type class `SelfDescribing`, we automatically can assume it can be `SelfDescribed`, but as I wrote in comment: this context-bound isn't necessary, I'm not planning to add it. its purpose was to make everything clearer... mission failed.

I'll add everything with tests and extensive comments.